File: deliverables/pri-35-carrossel-savi-feedmaker/generate_images.py
#!/usr/bin/env python3
import base64
import json
import os
import re
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path


ROOT = Path(__file__).resolve().parent
PROMPTS_PATH = ROOT / "image-prompts.md"
IMG_DIR = ROOT / "img"
API_URL = "https://api.openai.com/v1/images/generations"


def load_prompts():
    text = PROMPTS_PATH.read_text(encoding="utf-8")
    matches = re.findall(
        r"^## Slide\s+(\d{2})(?:[^\n]*)\n(.+?)(?=\n## Slide|\Z)",
        text,
        flags=re.M | re.S,
    )
    prompts = [(int(num), prompt.strip()) for num, prompt in matches]
    if len(prompts) != 9:
        raise RuntimeError(f"Expected 9 prompts, found {len(prompts)}")
    return prompts


def request_image(api_key, prompt, with_format=True):
    payload = {
        "model": "gpt-image-1",
        "prompt": prompt,
        "size": "1024x1536",
        "quality": "high",
        "n": 1,
    }
    if with_format:
        payload["output_format"] = "jpeg"
        payload["output_compression"] = 92

    body = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        API_URL,
        data=body,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=300) as response:
            return payload, json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as exc:
        details = exc.read().decode("utf-8", errors="replace")
        if with_format and "output_format" in details:
            return request_image(api_key, prompt, with_format=False)
        raise RuntimeError(f"OpenAI image request failed: HTTP {exc.code}: {details}") from exc


def main():
    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key:
        raise RuntimeError("OPENAI_API_KEY is not set")

    IMG_DIR.mkdir(parents=True, exist_ok=True)
    prompts = load_prompts()

    for index, prompt in prompts:
        out_path = IMG_DIR / f"slide_{index:02d}.jpg"
        if out_path.exists() and out_path.stat().st_size > 10_000:
            print(f"slide {index:02d}: exists, skipping", flush=True)
            continue

        print(f"slide {index:02d}: requesting image", flush=True)
        request_payload, result = request_image(api_key, prompt)
        b64 = result.get("data", [{}])[0].get("b64_json")
        if not b64:
            raise RuntimeError(f"slide {index:02d}: response did not include b64_json")

        out_path.write_bytes(base64.b64decode(b64))
        (IMG_DIR / f"request_{index:02d}.json").write_text(
            json.dumps(request_payload, ensure_ascii=False, indent=2),
            encoding="utf-8",
        )
        print(f"slide {index:02d}: saved {out_path} ({out_path.stat().st_size} bytes)", flush=True)
        time.sleep(1)

    print("image generation complete", flush=True)


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(str(exc), file=sys.stderr, flush=True)
        sys.exit(1)

File: deliverables/pri-60-carrossel-chatgpt-work/generate_images.js
const fs = require('fs');
const path = require('path');

const root = __dirname;
const promptsPath = path.join(root, 'image-prompts.md');
const imgDir = path.join(root, 'img');
const apiKey = process.env.GEMINI_API_KEY || process.env.GEMINI;
const apiUrl = 'https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash-image:generateContent';
const usedSlides = [1, 2, 5, 6, 7, 8];

function loadPrompts() {
  const text = fs.readFileSync(promptsPath, 'utf8');
  const matches = [...text.matchAll(/^## Slide\s+(\d{2})\n([\s\S]*?)(?=\n## Slide|\s*$)/gm)];
  const prompts = new Map();

  for (const match of matches) {
    prompts.set(Number(match[1]), match[2].trim());
  }

  return prompts;
}

async function requestImage(prompt) {
  const response = await fetch(apiUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-goog-api-key': apiKey,
    },
    body: JSON.stringify({
      contents: [{ parts: [{ text: prompt }] }],
      generationConfig: {
        responseModalities: ['IMAGE'],
        imageConfig: { aspectRatio: '4:5' },
      },
    }),
  });

  const data = await response.json();
  if (!response.ok) {
    throw new Error(`Gemini request failed: HTTP ${response.status}: ${JSON.stringify(data)}`);
  }

  const parts = data.candidates?.[0]?.content?.parts || [];
  const imagePart = parts.find((part) => part.inlineData?.data);
  if (!imagePart) {
    throw new Error(`Gemini response missing inline image data: ${JSON.stringify(data)}`);
  }

  return { data, base64: imagePart.inlineData.data };
}

async function main() {
  if (!apiKey) {
    throw new Error('GEMINI_API_KEY or GEMINI is not set');
  }

  fs.mkdirSync(imgDir, { recursive: true });
  const prompts = loadPrompts();

  for (const slide of usedSlides) {
    const prompt = prompts.get(slide);
    if (!prompt) {
      throw new Error(`Prompt missing for slide ${slide}`);
    }

    console.log(`slide ${String(slide).padStart(2, '0')}: requesting image`);
    const result = await requestImage(prompt);
    const outFile = path.join(imgDir, `slide_${String(slide).padStart(2, '0')}.jpg`);
    fs.writeFileSync(outFile, Buffer.from(result.base64, 'base64'));
    fs.writeFileSync(
      path.join(imgDir, `request_${String(slide).padStart(2, '0')}.json`),
      JSON.stringify({ model: 'gemini-2.5-flash-image', prompt }, null, 2),
      'utf8'
    );
    console.log(`slide ${String(slide).padStart(2, '0')}: saved ${outFile}`);
  }
}

main().catch((error) => {
  console.error(error.message || error);
  process.exit(1);
});
